SystemInformer/tokprp.c

```diff
@@ -3580,6 +3588,100 @@ PPH_STRING PhFormatTokenSecurityAttributeValue(
+    }
+
+    // APPID://SHA1HASH, APPID://SHA256HASH, APPID://SHA256FLATHASH
+    if (ValueIndex == 0 &&
+        Attribute->ValueType == TOKEN_SECURITY_ATTRIBUTE_TYPE_OCTET_STRING &&
+        ((Attribute->Values.OctetString[0].ValueLength == 20 && PhEqualStringRef(Name, &appidSHA1, TRUE)) ||
+        (Attribute->Values.OctetString[0].ValueLength == 32 && PhEqualStringRef(Name, &appidSHA256, TRUE)) ||
+        (Attribute->Values.OctetString[0].ValueLength == 32 && PhEqualStringRef(Name, &appidSHA256Flat, TRUE))))
+    {
+        static PWCHAR hexChars = L"0123456789ABCDEF";
+        PUCHAR hashValue = Attribute->Values.OctetString[0].Value;
+        PH_STRING_BUILDER sb;
+
+        PhInitializeStringBuilder(&sb, Attribute->Values.OctetString[0].ValueLength * 2 + sizeof(UNICODE_NULL));
+
+        for (ULONG i = 0; i < Attribute->Values.OctetString[0].ValueLength; i++)
+        {
+            PhAppendCharStringBuilder(&sb, hexChars[hashValue[i] >> 4]);
+            PhAppendCharStringBuilder(&sb, hexChars[hashValue[i] & 0x0F]);
+        }
+
+        return PhFinalStringBuilderString(&sb);
+    }
+
+    // SMARTLOCKER://SMARTSCREENORIGINCLAIM, SMARTLOCKER://SMARTSCREENORIGINCLAIMNOTINHERITED
+    if (ValueIndex == 0 &&
+        Attribute->ValueType == TOKEN_SECURITY_ATTRIBUTE_TYPE_OCTET_STRING &&
+        Attribute->Values.OctetString[0].ValueLength == sizeof(SE_SAFE_OPEN_PROMPT_RESULTS) &&
+        (PhEqualStringRef(Name, &originClaim, TRUE) || PhEqualStringRef(Name, &originClaimNI, TRUE)))
+    {
+        PSE_SAFE_OPEN_PROMPT_RESULTS claimValue = Attribute->Values.OctetString[0].Value;
+        PPH_STRING pathString;
+        PPH_STRING resultsString;
+        PPH_STRING string;
+
+#define PH_ORIGIN_CLAIM_RESULT(x, n) { TEXT(#x), x, FALSE, FALSE, n }
+
+        static const PH_ACCESS_ENTRY originClaimResults[] =
+        {
+            PH_ORIGIN_CLAIM_RESULT(SeSafeOpenExperienceCalled, L"Called"),
+            PH_ORIGIN_CLAIM_RESULT(SeSafeOpenExperienceAppRepCalled, L"App reputation called"),
+            PH_ORIGIN_CLAIM_RESULT(SeSafeOpenExperiencePromptDisplayed, L"Prompt displayed"),
+            PH_ORIGIN_CLAIM_RESULT(SeSafeOpenExperienceUAC, L"UAC"),
+            PH_ORIGIN_CLAIM_RESULT(SeSafeOpenExperienceUninstaller, L"Uninstaller"),
+            PH_ORIGIN_CLAIM_RESULT(SeSafeOpenExperienceIgnoreUnknownOrBad, L"Ignore unknown or bad"),
+            PH_ORIGIN_CLAIM_RESULT(SeSafeOpenExperienceDefenderTrustedInstaller, L"Defender trusted installer"),
+            PH_ORIGIN_CLAIM_RESULT(SeSafeOpenExperienceMOTWPresent, L"MOTW present"),
+            PH_ORIGIN_CLAIM_RESULT(SeSafeOpenExperienceElevatedNoPropagation, L"Elevated no propagation"),
+        };
+
+        resultsString = PhGetAccessString(claimValue->Results, (PPH_ACCESS_ENTRY)originClaimResults, RTL_NUMBER_OF(originClaimResults));
+        pathString = PhCreateStringEx(claimValue->Path, wcsnlen(claimValue->Path, sizeof(claimValue->Path) / sizeof(WCHAR)) * sizeof(WCHAR));
```

Comment:
It's probably worth us having a string creator to handle this case - constructing a zero terminated string up to a maximum length. Not requiring a change in PR here, but can be cleaned up later, suggesting:

```c
/**
 * Creates a string object up to a maximum length.
 *
 * \param Buffer A null-terminated Unicode string.
 * \param MaximumLength The maximum length, in bytes, of the string.
 */
FORCEINLINE
PPH_STRING
NTAPI
PhCreateStringZ2(
    _In_ PCWSTR String,
    _In_ ULONG MaximumLength
    )
{
    PH_STRINGREF string;

    string.Length = wcsnlen(String, MaximumLength / sizeof(WCHAR)) * sizeof(WCHAR);
    string.Buffer = (PWSTR)String;

    return PhCreateString2(&string);
}
```